File: src/ibkr_porez/declaration_income_xml.py
"""Generator for PP OPO (Capital Income) XML declarations."""

from datetime import date, timedelta
from decimal import ROUND_HALF_UP, Decimal
from xml.dom import minidom

import holidays

from ibkr_porez.models import (
    INCOME_CODE_COUPON,
    INCOME_CODE_DIVIDEND,
    IncomeEntry,
    UserConfig,
)


class IncomeXMLGenerator:
    """Generator for PP OPO XML declarations."""

    # Tax declaration due date: 30 days after declaration date (legal requirement)
    TAX_DUE_DATE_DAYS = 30

    def __init__(self, config: UserConfig):
        self.config = config

    def generate_xml(  # noqa: PLR0915
        self,
        income_entries: list[IncomeEntry],
        declaration_date: date,
        income_type: str,  # "dividend" or "coupon"
        withholding_tax_rsd: Decimal = Decimal("0.00"),
    ) -> str:
        """
        Generate PP OPO XML for given income entries.

        Args:
            income_entries: List of income entries (all should be same date and type).
            declaration_date: Date of income realization (YYYY-MM-DD).
            income_type: Type of income ("dividend" or "coupon").

        Returns:
            str: XML content for PP OPO declaration.
        """
        doc = minidom.Document()

        # Root Element: ns1:PodaciPoreskeDeklaracije
        root = doc.createElement("ns1:PodaciPoreskeDeklaracije")
        root.setAttribute("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance")
        root.setAttribute("xmlns:ns1", "http://pid.purs.gov.rs")
        doc.appendChild(root)

        def create_text(parent, tag, value):
            el = doc.createElement(f"ns1:{tag}")
            if value is not None:
                el.appendChild(doc.createTextNode(str(value)))
            parent.appendChild(el)

        def create_cdata(parent, tag, value):
            el = doc.createElement(f"ns1:{tag}")
            cdata = doc.createCDATASection(str(value) if value else "")
            el.appendChild(cdata)
            parent.appendChild(el)

        p_prijavi = doc.createElement("ns1:PodaciOPrijavi")
        root.appendChild(p_prijavi)

        create_text(p_prijavi, "VrstaPrijave", "1")
        period_str = declaration_date.strftime("%Y-%m")
        create_text(p_prijavi, "ObracunskiPeriod", period_str)
        create_text(p_prijavi, "DatumOstvarivanjaPrihoda", declaration_date.strftime("%Y-%m-%d"))
        create_text(p_prijavi, "Rok", "1")
        # If weekend/holiday -> first next working day
        saturday = 5
        base_due = declaration_date + timedelta(days=self.TAX_DUE_DATE_DAYS)
        rs_holidays = holidays.country_holidays("RS")

        # Shift if weekend (5=Sat, 6=Sun) or Holiday
        while base_due.weekday() >= saturday or base_due in rs_holidays:
            base_due += timedelta(days=1)

        create_text(p_prijavi, "DatumDospelostiObaveze", base_due.strftime("%Y-%m-%d"))

        p_obveznik = doc.createElement("ns1:PodaciOPoreskomObvezniku")
        root.appendChild(p_obveznik)

        create_text(p_obveznik, "PoreskiIdentifikacioniBroj", self.config.personal_id)
        create_cdata(p_obveznik, "ImePrezimeObveznika", self.config.full_name)
        create_cdata(p_obveznik, "UlicaBrojPoreskogObveznika", self.config.address)
        create_text(p_obveznik, "PrebivalisteOpstina", self.config.city_code)
        create_text(p_obveznik, "JMBGPodnosiocaPrijave", self.config.personal_id)
        create_text(p_obveznik, "TelefonKontaktOsobe", self.config.phone)
        create_cdata(p_obveznik, "ElektronskaPosta", self.config.email)

        p_nacin = doc.createElement("ns1:PodaciONacinuOstvarivanjaPrihoda")
        root.appendChild(p_nacin)

        create_text(p_nacin, "NacinIsplate", "3")
        create_text(p_nacin, "Ostalo", "Isplata na brokerski racun")

        deklaracija = doc.createElement("ns1:DeklarisaniPodaciOVrstamaPrihoda")
        root.appendChild(deklaracija)

        # Calculate totals
        total_bruto = sum(entry.amount_rsd for entry in income_entries)
        # Round to 2 decimal places
        total_bruto = round(total_bruto, 2)

        # Tax rate: 15% for capital income
        tax_rate = Decimal("0.15")
        osnovica = total_bruto
        # Use quantize with ROUND_HALF_UP for taxes (standard rounding)

        obracunati_porez = (osnovica * tax_rate).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

        # Calculate foreign tax paid (withholding tax)
        # This comes from WITHHOLDING_TAX transactions, passed as parameter
        porez_placen_drugoj_drzavi = withholding_tax_rsd

        # Round withholding tax
        porez_placen_drugoj_drzavi = round(porez_placen_drugoj_drzavi, 2)

        # PorezZaUplatu = ObracunatiPorez - PorezPlacenDrugojDrzavi
        porez_za_uplatu = max(Decimal("0.00"), obracunati_porez - porez_placen_drugoj_drzavi)
        porez_za_uplatu = round(porez_za_uplatu, 2)

        sifra_vrste_prihoda = (
            INCOME_CODE_DIVIDEND if income_type == "dividend" else INCOME_CODE_COUPON
        )

        podaci_vrsta = doc.createElement("ns1:PodaciOVrstamaPrihoda")
        deklaracija.appendChild(podaci_vrsta)

        create_text(podaci_vrsta, "RedniBroj", "1")
        create_text(podaci_vrsta, "SifraVrstePrihoda", sifra_vrste_prihoda)
        create_text(podaci_vrsta, "BrutoPrihod", f"{total_bruto:.2f}")
        create_text(podaci_vrsta, "OsnovicaZaPorez", f"{osnovica:.2f}")
        create_text(podaci_vrsta, "ObracunatiPorez", f"{obracunati_porez:.2f}")
        create_text(podaci_vrsta, "PorezPlacenDrugojDrzavi", f"{porez_placen_drugoj_drzavi:.2f}")
        create_text(podaci_vrsta, "PorezZaUplatu", f"{porez_za_uplatu:.2f}")

        ukupno = doc.createElement("ns1:Ukupno")
        root.appendChild(ukupno)

        create_text(ukupno, "FondSati", "0.00")
        create_text(ukupno, "BrutoPrihod", f"{total_bruto:.2f}")
        create_text(ukupno, "OsnovicaZaPorez", f"{osnovica:.2f}")
        create_text(ukupno, "ObracunatiPorez", f"{obracunati_porez:.2f}")
        create_text(ukupno, "PorezPlacenDrugojDrzavi", f"{porez_placen_drugoj_drzavi:.2f}")
        create_text(ukupno, "PorezZaUplatu", f"{porez_za_uplatu:.2f}")
        create_text(ukupno, "OsnovicaZaDoprinose", "0.00")
        create_text(ukupno, "PIO", "0.00")
        create_text(ukupno, "ZDRAVSTVO", "0.00")
        create_text(ukupno, "NEZAPOSLENOST", "0.00")

        kamata = doc.createElement("ns1:Kamata")
        root.appendChild(kamata)

        create_text(kamata, "PorezZaUplatu", "0")
        create_text(kamata, "OsnovicaZaDoprinose", "0")
        create_text(kamata, "PIO", "0")
        create_text(kamata, "ZDRAVSTVO", "0")
        create_text(kamata, "NEZAPOSLENOST", "0")

        dodatna_kamata = doc.createElement("ns1:PodaciODodatnojKamati")
        root.appendChild(dodatna_kamata)

        return doc.toprettyxml(indent="  ")

File: src/declaration_income_xml.rs
use chrono::NaiveDate;
use quick_xml::Writer;
use quick_xml::events::{BytesCData, BytesEnd, BytesStart, BytesText, Event};
use rust_decimal::Decimal;

use crate::due_date::next_working_day;
use crate::holidays::HolidayCalendar;
use crate::models::{IncomeDeclarationEntry, UserConfig};

#[must_use]
#[allow(clippy::missing_panics_doc)]
pub fn generate_income_xml(
    entry: &IncomeDeclarationEntry,
    config: &UserConfig,
    holidays: &HolidayCalendar,
) -> String {
    let due_date = next_working_day(entry.date, holidays);

    let mut buf = Vec::new();
    let mut w = Writer::new_with_indent(&mut buf, b' ', 2);

    w.write_event(Event::Decl(quick_xml::events::BytesDecl::new(
        "1.0", None, None,
    )))
    .unwrap();

    let mut root = BytesStart::new("ns1:PodaciPoreskeDeklaracije");
    root.push_attribute(("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance"));
    root.push_attribute(("xmlns:ns1", "http://pid.purs.gov.rs"));
    w.write_event(Event::Start(root)).unwrap();

    write_podaci_o_prijavi(&mut w, entry, due_date);
    write_poreski_obveznik(&mut w, config);
    write_nacin_ostvarivanja(&mut w);
    write_vrste_prihoda(&mut w, entry);
    write_ukupno(&mut w, entry);
    write_kamata(&mut w);
    write_dodatna_kamata(&mut w);

    w.write_event(Event::End(BytesEnd::new("ns1:PodaciPoreskeDeklaracije")))
        .unwrap();

    String::from_utf8(buf).unwrap()
}

fn write_podaci_o_prijavi(
    w: &mut Writer<&mut Vec<u8>>,
    entry: &IncomeDeclarationEntry,
    due_date: NaiveDate,
) {
    start(w, "ns1:PodaciOPrijavi");
    text_elem(w, "ns1:VrstaPrijave", "1");
    text_elem(
        w,
        "ns1:ObracunskiPeriod",
        &entry.date.format("%Y-%m").to_string(),
    );
    text_elem(w, "ns1:DatumOstvarivanjaPrihoda", &fmt_date(entry.date));
    text_elem(w, "ns1:Rok", "1");
    text_elem(w, "ns1:DatumDospelostiObaveze", &fmt_date(due_date));
    end(w, "ns1:PodaciOPrijavi");
}

fn write_poreski_obveznik(w: &mut Writer<&mut Vec<u8>>, config: &UserConfig) {
    start(w, "ns1:PodaciOPoreskomObvezniku");
    text_elem(w, "ns1:PoreskiIdentifikacioniBroj", &config.personal_id);
    cdata_elem(w, "ns1:ImePrezimeObveznika", &config.full_name);
    cdata_elem(w, "ns1:UlicaBrojPoreskogObveznika", &config.address);
    text_elem(w, "ns1:PrebivalisteOpstina", &config.city_code);
    text_elem(w, "ns1:JMBGPodnosiocaPrijave", &config.personal_id);
    text_elem(w, "ns1:TelefonKontaktOsobe", &config.phone);
    cdata_elem(w, "ns1:ElektronskaPosta", &config.email);
    end(w, "ns1:PodaciOPoreskomObvezniku");
}

fn write_nacin_ostvarivanja(w: &mut Writer<&mut Vec<u8>>) {
    start(w, "ns1:PodaciONacinuOstvarivanjaPrihoda");
    text_elem(w, "ns1:NacinIsplate", "3");
    text_elem(w, "ns1:Ostalo", "Isplata na brokerski racun");
    end(w, "ns1:PodaciONacinuOstvarivanjaPrihoda");
}

fn write_vrste_prihoda(w: &mut Writer<&mut Vec<u8>>, entry: &IncomeDeclarationEntry) {
    start(w, "ns1:DeklarisaniPodaciOVrstamaPrihoda");
    start(w, "ns1:PodaciOVrstamaPrihoda");
    text_elem(w, "ns1:RedniBroj", "1");
    text_elem(w, "ns1:SifraVrstePrihoda", &entry.sifra_vrste_prihoda);
    text_elem(w, "ns1:BrutoPrihod", &fmt2(entry.bruto_prihod));
    text_elem(w, "ns1:OsnovicaZaPorez", &fmt2(entry.osnovica_za_porez));
    text_elem(w, "ns1:ObracunatiPorez", &fmt2(entry.obracunati_porez));
    text_elem(
        w,
        "ns1:PorezPlacenDrugojDrzavi",
        &fmt2(entry.porez_placen_drugoj_drzavi),
    );
    text_elem(w, "ns1:PorezZaUplatu", &fmt2(entry.porez_za_uplatu));
    end(w, "ns1:PodaciOVrstamaPrihoda");
    end(w, "ns1:DeklarisaniPodaciOVrstamaPrihoda");
}

fn write_ukupno(w: &mut Writer<&mut Vec<u8>>, entry: &IncomeDeclarationEntry) {
    start(w, "ns1:Ukupno");
    text_elem(w, "ns1:FondSati", "0.00");
    text_elem(w, "ns1:BrutoPrihod", &fmt2(entry.bruto_prihod));
    text_elem(w, "ns1:OsnovicaZaPorez", &fmt2(entry.osnovica_za_porez));
    text_elem(w, "ns1:ObracunatiPorez", &fmt2(entry.obracunati_porez));
    text_elem(
        w,
        "ns1:PorezPlacenDrugojDrzavi",
        &fmt2(entry.porez_placen_drugoj_drzavi),
    );
    text_elem(w, "ns1:PorezZaUplatu", &fmt2(entry.porez_za_uplatu));
    text_elem(w, "ns1:OsnovicaZaDoprinose", "0.00");
    text_elem(w, "ns1:PIO", "0.00");
    text_elem(w, "ns1:ZDRAVSTVO", "0.00");
    text_elem(w, "ns1:NEZAPOSLENOST", "0.00");
    end(w, "ns1:Ukupno");
}

fn write_kamata(w: &mut Writer<&mut Vec<u8>>) {
    start(w, "ns1:Kamata");
    text_elem(w, "ns1:PorezZaUplatu", "0");
    text_elem(w, "ns1:OsnovicaZaDoprinose", "0");
    text_elem(w, "ns1:PIO", "0");
    text_elem(w, "ns1:ZDRAVSTVO", "0");
    text_elem(w, "ns1:NEZAPOSLENOST", "0");
    end(w, "ns1:Kamata");
}

fn write_dodatna_kamata(w: &mut Writer<&mut Vec<u8>>) {
    start(w, "ns1:PodaciODodatnojKamati");
    end(w, "ns1:PodaciODodatnojKamati");
}

fn start(w: &mut Writer<&mut Vec<u8>>, name: &str) {
    w.write_event(Event::Start(BytesStart::new(name))).unwrap();
}

fn end(w: &mut Writer<&mut Vec<u8>>, name: &str) {
    w.write_event(Event::End(BytesEnd::new(name))).unwrap();
}

fn text_elem(w: &mut Writer<&mut Vec<u8>>, name: &str, value: &str) {
    w.write_event(Event::Start(BytesStart::new(name))).unwrap();
    w.write_event(Event::Text(BytesText::new(value))).unwrap();
    w.write_event(Event::End(BytesEnd::new(name))).unwrap();
}

fn cdata_elem(w: &mut Writer<&mut Vec<u8>>, name: &str, value: &str) {
    w.write_event(Event::Start(BytesStart::new(name))).unwrap();
    w.write_event(Event::CData(BytesCData::new(value))).unwrap();
    w.write_event(Event::End(BytesEnd::new(name))).unwrap();
}

fn fmt_date(d: NaiveDate) -> String {
    d.format("%Y-%m-%d").to_string()
}

fn fmt2(d: Decimal) -> String {
    format!("{d:.2}")
}
